src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -3,311 +3,820 @@
+                string code = item.GetMetadata(MetadataKeys.DiagnosticCode);
+
+                switch (severity)
+                {
+                    case nameof(LogLevel.Error):
+                        Log.LogError(null, code, null, ProjectPath, 0, 0, 0, 0, message);
+                        break;
+                    case nameof(LogLevel.Warning):
+                        Log.LogWarning(null, code, null, ProjectPath, 0, 0, 0, 0, message);
+                        break;
+                    default:
+                        Log.LogMessage(null, code, null, ProjectPath, 0, 0, 0, 0, message);
+                        break;
+                }
             }
+        }
 
-            ContentFilesToPreprocess = RaisePackageAssets(
-                runtimeTarget,
-                p => p.ContentFiles,
-                filter: asset => !string.IsNullOrEmpty(asset.PPOutputPath),
-                setup: (asset, item) => 
+        private byte[] HashSettings()
+        {
+            using (var stream = new MemoryStream())
+            {
+                using (var writer = new BinaryWriter(stream, TextEncoding, leaveOpen: true))
                 {
-                    item.SetMetadata(MetadataKeys.BuildAction, asset.BuildAction.ToString());
-                    item.SetMetadata(MetadataKeys.CopyToOutput, asset.CopyToOutput.ToString());
-                    item.SetMetadata(MetadataKeys.PPOutputPath, asset.PPOutputPath);
+                    writer.Write(ProjectPath);
+                    writer.Write(TargetFrameworkMoniker);
+                    writer.Write(DisableTransitiveProjectReferences);
+                    writer.Write(DisableFrameworkAssemblies);
+                    writer.Write(MarkPackageReferencesAsExternallyResolved);
+                    writer.Write(ProjectLanguage ?? "");
+                    writer.Write(RuntimeIdentifier ?? "");
+                }
+
+                stream.Position = 0;
 
-                    if (!string.IsNullOrEmpty(MetadataKeys.OutputPath))
+                using (var hash = CreateSettingsHash())
+                {
+                    return hash.ComputeHash(stream);
+                }
+            }
+        }
+
+        private sealed class CacheReader : IDisposable
+        {
+            private BinaryReader _reader;
+            private string[] _metadataStringTable;
+
+            public CacheReader(ResolvePackageAssets task)
+            {
+                byte[] settingsHash = task.HashSettings();
+                BinaryReader reader = null;
+
+                try
+                {
+                    if (File.GetLastWriteTimeUtc(task.ProjectAssetsCacheFile) >= File.GetLastWriteTimeUtc(task.ProjectAssetsFile))
                     {
-                        item.SetMetadata(MetadataKeys.OutputPath, asset.OutputPath);
+                        reader = OpenCacheFile(task.ProjectAssetsCacheFile, settingsHash);
                     }
+                }
```

Comment:
It doesn't look like you handle InvalidDataException here - I would have thought we'd want to throw away the cache in that case?

Reply:
InvalidDataException doesn't derive from IOException - which you may have thought it did.

Reply:
Good catch. Need to add tests for this. I did think it derived from IOException.

Reply:
Fixed and test coverage added.